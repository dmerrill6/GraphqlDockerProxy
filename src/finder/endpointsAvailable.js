import fetch from 'node-fetch'
import type { Endpoints } from './findEndpointsInterface'
import sleep from 'sleep-promise'
const queryStr = '?query=%7B__schema%7Btypes%7Bname%7D%7D%7D'


export const allEndpointsAvailable = async(endpoints :Endpoints): Endpoints => {
  winston.info('check that all endpoints are available')
  let allEndpointsAvailable = false
  let counter = 0
  while (!allEndpointsAvailable && counter < 10){
    allEndpointsAvailable = true
    counter = counter + 1
    for (const namespace in endpoints){
      const oneNamespace = endpoints[namespace]
      for (let i = 0; i < oneNamespace.length; i++){
        const one = oneNamespace[i]
        const result = await fetch(one.url + queryStr, {
          timeout: 2000,
        }).then(res => {
          return {
            status: res.status,
            ok: res.ok,
          }
        })
          .catch(() => {
            return {
              status: 404,
              ok: false,
            }
          })
        if (!result.ok){
          winston.warn('allEndpointsAvailable: Endpoint not Available: ', one)
          allEndpointsAvailable = false
        }

      }
    }
    if (!allEndpointsAvailable){
      winston.info('wait a second and try it again')
      await sleep(1000)
    }
  }
  if (!allEndpointsAvailable){
    winston.warn('some Endpoint not Available')
    return false
  }
  return true
}

export const sortEndpointAndFindAvailableEndpoints = async(endpoints :Endpoints): Endpoints => {
  for (const namespace in endpoints){
    const oneNamespace = endpoints[namespace]
    for (let i = 0; i < oneNamespace.length; i++){
      const one = oneNamespace[i]
      const result = await fetch(one.url + queryStr, {
        timeout: 2000,
      }).then(res => {
        return {
          status: res.status,
          ok: res.ok,
        }
      })
        .catch(() => {
          return {
            status: 404,
            ok: false,
          }
        })
      if (!result.ok){
        winston.warn('Endpoint not Available: ', one)
        delete oneNamespace.splice(i, 1)
      }

    }
    if (oneNamespace.length === 0){
      delete endpoints[namespace]
    }

  }
  return endpoints
}
