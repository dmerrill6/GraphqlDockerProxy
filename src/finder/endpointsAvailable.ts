import fetch from 'node-fetch';
import  { Endpoints } from './findEndpointsInterface';
const queryStr = '?query=%7B__schema%7Btypes%7Bname%7D%7D%7D';

export const sortEndpointAndFindAvailableEndpoints =
async(endpoints :Endpoints): Promise<Endpoints> => {
  for (const namespace in endpoints) {
    const oneNamespace = endpoints[namespace];
    for (let i = 0; i < oneNamespace.length; i = + 1) {
      const one = oneNamespace[i];
      const result = await fetch(one.url + queryStr, {
        timeout: 2000,
      }).then((res) => {
        return {
          status: res.status,
          ok: res.ok,
        };
      })
        .catch(() => {
          return {
            status: 404,
            ok: false,
          };
        });
      if (!result.ok) {
        winston.warn('Endpoint not Available: ', one);
        oneNamespace.splice(i, 1);
      }

    }
    if (oneNamespace.length === 0) {
      delete endpoints[namespace];
    }

  }
  return endpoints;
};
