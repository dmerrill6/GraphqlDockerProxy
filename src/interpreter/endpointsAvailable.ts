import fetch from 'node-fetch'
import  { Endpoints } from './endpoints'
// tslint:disable-next-line
const queryStr = '?query=%0A%20%20%20%20query%20IntrospectionQuery%20%7B%0A%20%20%20%20%20%20__schema%20%7B%0A%20%20%20%20%20%20%20%20queryType%20%7B%20name%20%7D%0A%20%20%20%20%20%20%20%20mutationType%20%7B%20name%20%7D%0A%20%20%20%20%20%20%20%20subscriptionType%20%7B%20name%20%7D%0A%20%20%20%20%20%20%20%20types%20%7B%0A%20%20%20%20%20%20%20%20%20%20...FullType%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20directives%20%7B%0A%20%20%20%20%20%20%20%20%20%20name%0A%20%20%20%20%20%20%20%20%20%20description%0A%20%20%20%20%20%20%20%20%20%20locations%0A%20%20%20%20%20%20%20%20%20%20args%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20...InputValue%0A%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%0A%0A%20%20%20%20fragment%20FullType%20on%20__Type%20%7B%0A%20%20%20%20%20%20kind%0A%20%20%20%20%20%20name%0A%20%20%20%20%20%20description%0A%20%20%20%20%20%20fields(includeDeprecated%3A%20true)%20%7B%0A%20%20%20%20%20%20%20%20name%0A%20%20%20%20%20%20%20%20description%0A%20%20%20%20%20%20%20%20args%20%7B%0A%20%20%20%20%20%20%20%20%20%20...InputValue%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20type%20%7B%0A%20%20%20%20%20%20%20%20%20%20...TypeRef%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20isDeprecated%0A%20%20%20%20%20%20%20%20deprecationReason%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20inputFields%20%7B%0A%20%20%20%20%20%20%20%20...InputValue%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20interfaces%20%7B%0A%20%20%20%20%20%20%20%20...TypeRef%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20enumValues(includeDeprecated%3A%20true)%20%7B%0A%20%20%20%20%20%20%20%20name%0A%20%20%20%20%20%20%20%20description%0A%20%20%20%20%20%20%20%20isDeprecated%0A%20%20%20%20%20%20%20%20deprecationReason%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20possibleTypes%20%7B%0A%20%20%20%20%20%20%20%20...TypeRef%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%0A%0A%20%20%20%20fragment%20InputValue%20on%20__InputValue%20%7B%0A%20%20%20%20%20%20name%0A%20%20%20%20%20%20description%0A%20%20%20%20%20%20type%20%7B%20...TypeRef%20%7D%0A%20%20%20%20%20%20defaultValue%0A%20%20%20%20%7D%0A%0A%20%20%20%20fragment%20TypeRef%20on%20__Type%20%7B%0A%20%20%20%20%20%20kind%0A%20%20%20%20%20%20name%0A%20%20%20%20%20%20ofType%20%7B%0A%20%20%20%20%20%20%20%20kind%0A%20%20%20%20%20%20%20%20name%0A%20%20%20%20%20%20%20%20ofType%20%7B%0A%20%20%20%20%20%20%20%20%20%20kind%0A%20%20%20%20%20%20%20%20%20%20name%0A%20%20%20%20%20%20%20%20%20%20ofType%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20kind%0A%20%20%20%20%20%20%20%20%20%20%20%20name%0A%20%20%20%20%20%20%20%20%20%20%20%20ofType%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20kind%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20name%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20ofType%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20kind%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20name%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20ofType%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20kind%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20name%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20ofType%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20kind%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20name%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%0A%20%20' 
export const sortEndpointAndFindAvailableEndpoints =
async(endpoints :Endpoints): Promise<Endpoints> => {
  for (const namespace in endpoints) {
    const oneNamespace = endpoints[namespace]
    for (let i = 0; i < oneNamespace.length; i = i + 1) {
      const one = oneNamespace[i]
      const result = await fetch(one.url + queryStr, {
        timeout: 2000,
      }).then(async(res) => {
        return {
          json: await res.json(),
          status: res.status,
          ok: res.ok,
        }
      })
        .catch((err) => {
          winston.debug('Error by fetching', { err, url: one.url })
          return {
            json: {},
            status: 404,
            ok: false,
          }
        })
      if (!result.ok) {
        winston.warn('Endpoint not Available: ', { url: one.url , namespace: one.namespace })
        oneNamespace.splice(i, 1)
      } else {
        oneNamespace[i].__introspection = result.json
      }

    }
    if (oneNamespace.length === 0) {
      delete endpoints[namespace]
    }

  }
  return endpoints
}
