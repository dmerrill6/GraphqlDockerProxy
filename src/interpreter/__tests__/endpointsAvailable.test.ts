import { sortEndpointAndFindAvailableEndpoints } from '../endpointsAvailable'

import { Endpoints } from '../endpoints'
jest.mock('node-fetch', () => {
  return {
    default: (url) => {
      if (url.startsWith('found')) {
        // found
        return Promise.resolve({
          ok: true,
          status: 200,
          json: () => {
            return Promise.resolve({
              found: true,
            })
          },
        })
      }
      return Promise.resolve({
        ok: false,
        status: 200,
        json: () => {
          return Promise.resolve({
            found: false,
          })
        },
      })

    },
  }
})
describe('test endpointsAvailable', () => {
  it('tests that not available endpoints are filtert', async() => {
    const testEndpoints : Endpoints = {
      b: [
        {
          url: 'found',
          namespace: 'mock1',
          typePrefix: '_mock1',
          __imageID: '___',
          __deploymentName: '___',

        },
        {
          url: 'notfound',
          namespace: 'mock1',
          typePrefix: '_mock1',
          __imageID: '___',
          __deploymentName: '___',

        },
      ],
      a: [
        {
          url: 'found',
          namespace: 'mock1',
          typePrefix: '_mock1',
          __imageID: '___',
          __deploymentName: '___',

        },
        {
          url: 'found',
          namespace: 'mock1',
          typePrefix: '_mock1',
          __imageID: '___',
          __deploymentName: '___',

        },
      ],
    }

    const result = await sortEndpointAndFindAvailableEndpoints(testEndpoints)
    expect(result.a.length).toBe(2)
    expect(result.b.length).toBe(1)
    expect(result).toMatchSnapshot()
  })
})
