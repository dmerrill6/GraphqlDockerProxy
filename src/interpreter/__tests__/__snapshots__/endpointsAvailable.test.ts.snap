// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test endpointsAvailable tests that not available endpoints are filtert 1`] = `
Object {
  "a": Array [
    Object {
      "__deploymentName": "___",
      "__imageID": "___",
      "__introspection": Object {
        "found": true,
      },
      "namespace": "mock1",
      "typePrefix": "_mock1",
      "url": "found",
    },
    Object {
      "__deploymentName": "___",
      "__imageID": "___",
      "__introspection": Object {
        "found": true,
      },
      "namespace": "mock1",
      "typePrefix": "_mock1",
      "url": "found",
    },
  ],
  "b": Array [
    Object {
      "__deploymentName": "___",
      "__imageID": "___",
      "__introspection": Object {
        "found": true,
      },
      "namespace": "mock1",
      "typePrefix": "_mock1",
      "url": "found",
    },
  ],
}
`;
